### PR TITLE
Federal account endpoints

### PR DESCRIPTION
**Pull request by afrasier**
Changes:

* Added `/federal_accounts/` which will list/retrieve federal account objects
* Added `/federal_accounts/autocomplete` to support autocomplete on federal accounts
* Moved `/accounts/tas/` (And related endpoints `. . balances/` `. . categories/` and `totals`) to `/tas/` ( shout outs to @mab6bf @kevinli-work )
  - `/accounts/tas/` -> `/tas/`
  - `/accounts/tas/autocomplete/` -> `/tas/autocomplete`
  - `/accounts/tas/balances/` -> `/tas/balances/`
  - `/accounts/tas/balances/total/` -> `/tas/balances/total`
  - `/accounts/tas/categories/` -> `/tas/categories/`
  - `/accounts/tas/categories/total` -> `/tas/categories/total`
* Updated documentation for the new endpoints, and refactored endpoint paths
* Added tests for new endpoints, updated tests for refactored endpoints
* Refactored `/accounts/views.py` into 3 files `common.py`, `tas.py`, `federal.py` under the `/accounts/views/` folder
* Refactored urls in `/accounts/urls.py` to `urls.py`, `urls_federal.py` and `urls_tas.py`

### CODE REVIEW

File: usaspending_api/accounts/tests/test_federal_account_autocomplete.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        account_title='zzz',
+        main_account_code='abc')
+    mommy.make(
+        FederalAccount,
+        account_title='###',
+        main_account_code='789')
+    mommy.make(FederalAccount, main_account_code='XYZ')
+
+
+@pytest.mark.parametrize("fields,value,expected", [
```
Comment (bsweger) [inline usaspending_api/accounts/tests/test_federal_account_autocomplete.py]: You've been doing them all along, but just wanted to say again how much I love these parameterized tests courtesy of pytest

File: usaspending_api/accounts/tests/test_federal_endpoints.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import pytest
```
Comment (bsweger) [inline usaspending_api/accounts/tests/test_federal_endpoints.py]: Am leaving a few comments throughout the PR re: the name `federal` instead of `federal_account`. In the long run, I'm wondering if the shorter names are worth the potential ambiguity/confusion. This is strictly backend stuff, so no need hold up the merge, but what do you think about using `test_federal_account_endpoints.py` as the file name here?

File: usaspending_api/accounts/urls_federal.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
+from django.conf.urls import url
```
Comment (bsweger) [inline usaspending_api/accounts/urls_federal.py]: Same question here about `urls_federal_account.py` instead of `ursl_federal.py`

File: usaspending_api/accounts/views/federal.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
+from usaspending_api.accounts.serializers import FederalAccountSerializer
```
Comment (bsweger) [inline usaspending_api/accounts/views/federal.py]: Thanks for splitting out federal accounts and TAS! I have the same question about calling this file `federal_account.py` instead of `federal.py`

File: usaspending_api/accounts/views/common.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
+from usaspending_api.awards.serializers import FinancialAccountsByAwardsSerializer
```
Comment (bsweger) [inline usaspending_api/accounts/views/common.py]: Poor File C data, we never know where to put it. I figured out that one reason it's hard to find this in the code base is that the model for it lives in the `awards` application and the view for it lives in the `accounts` application. 

If we had to pick one, I'd go with `accounts`, like you have it here (because that data ultimately can only come from an agency DATA Act submission and not from current USAspending info). Just marking your cards that the next time we have a chance, we might want to move the model over to `accounts` so they can live together.

Lastly, since this data is keyed by TAS, what do you think about moving this single view into `tas.py` and getting rid of `common.py`?

File: usaspending_api/api_docs/markdown/data_dictionary.md [hunk 6]
```diff
@@ -98,12 +98,14 @@ In this section you will find definitions for common terms used in the API. For
...
-| [/api/v1/accounts/categories/](/api/v1/accounts/categories/) | GET, POST | <a href="#accounts-prg-obj">Appropriation Account Balances (by Category)</a> | Returns a list of appropriation account balances broken up by program activities and object class |
-| [/api/v1/accounts/categories/total/](/api/v1/accounts/categories/total/) | POST |  Aggregate (see [Using the API](/docs/using-the-api)) | Supports aggregation on appropriation account (by category) records |
-| [/api/v1/accounts/tas/](/api/v1/accounts/tas/) | GET, POST | <a href="#tas">Treasury Appropriation Account</a> | Returns a list of treasury appropriation accounts, by TAS |
-| [/api/v1/accounts/tas/autocomplete/](/api/v1/accounts/tas/autocomplete/) | POST | Autocomplete (see [Using the API](/docs/using-the-api))| Supports autocomplete on TAS records |
+| [/api/v1/federal_accounts/](/api/v1/federal_accounts/) | GET, POST | <a href="#federal-account">Federal Account</a> | Returns a list of federal accounts |
+| [/api/v1/federal_accounts/total/](/api/v1/federal_accounts/total/) | POST |  Aggregate (see [Using the API](/docs/using-the-api)) | Supports aggregation on federal accounts records |
```
Comment (bsweger) [inline usaspending_api/api_docs/markdown/data_dictionary.md]: Since there isn't any numeric data on `federal_accounts`, do we need a `/total` endpoint for it?

File: usaspending_api/api_docs/markdown/data_dictionary.md [hunk 7]
```diff
@@ -98,12 +98,14 @@ In this section you will find definitions for common terms used in the API. For
...
-| [/api/v1/accounts/balances/total/](/api/v1/accounts/balances/total/) | POST |  Aggregate (see [Using the API](/docs/using-the-api)) | Supports aggregation on appropriation account records |
-| [/api/v1/accounts/categories/](/api/v1/accounts/categories/) | GET, POST | <a href="#accounts-prg-obj">Appropriation Account Balances (by Category)</a> | Returns a list of appropriation account balances broken up by program activities and object class |
-| [/api/v1/accounts/categories/total/](/api/v1/accounts/categories/total/) | POST |  Aggregate (see [Using the API](/docs/using-the-api)) | Supports aggregation on appropriation account (by category) records |
-| [/api/v1/accounts/tas/](/api/v1/accounts/tas/) | GET, POST | <a href="#tas">Treasury Appropriation Account</a> | Returns a list of treasury appropriation accounts, by TAS |
-| [/api/v1/accounts/tas/autocomplete/](/api/v1/accounts/tas/autocomplete/) | POST | Autocomplete (see [Using the API](/docs/using-the-api))| Supports autocomplete on TAS records |
+| [/api/v1/federal_accounts/](/api/v1/federal_accounts/) | GET, POST | <a href="#federal-account">Federal Account</a> | Returns a list of federal accounts |
```
Comment (bsweger) [inline usaspending_api/api_docs/markdown/data_dictionary.md]: As always, thanks for being a diligent doc updater

File: usaspending_api/accounts/tests/test_federal_endpoints.py [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+from model_mommy import mommy
+import json
+
+
+@pytest.fixture
+def federal_account_models():
+    fed_1 = mommy.make('accounts.FederalAccount', account_title="Community Spending", _fill_optional=True)
+    fed_2 = mommy.make('accounts.FederalAccount', account_title="Military Spending", _fill_optional=True)
+    fed_3 = mommy.make('accounts.FederalAccount', account_title="Foreign Spending", _fill_optional=True)
+    tas_1 = mommy.make('accounts.TreasuryAppropriationAccount', federal_account=fed_1, tas_rendering_label="ABC", _fill_optional=True)
```
Comment (bsweger) [inline usaspending_api/accounts/tests/test_federal_endpoints.py]: Do you need the `tas_1` and `tas_2` models for these tests?

File: usaspending_api/accounts/urls.py [hunk 9]
```diff
@@ -1,27 +1,11 @@
 from django.conf.urls import url
```
Comment (bsweger) [inline usaspending_api/accounts/urls.py]: Similar question that I had about common.py in `views/` (below). Since this File C data is the only thing here and is keyed by tas, what do you think about moving these urls to `tas.py` and getting rid of `common.py` since it's not really "common" as much as it is "orphaned."

File: usaspending_api/data/testing_data/endpoint_testing_data.json [hunk 10]
```diff
@@ -525,7 +525,7 @@
             ]
```
Comment (bsweger) [inline usaspending_api/data/testing_data/endpoint_testing_data.json]: I'm curious about what we think the endpoint tests should be doing these days. We're definitely better about using unit tests to hit edge cases. Are these still useful as a high-level integration test? Do need to add the new endpoints here? Or is this file strictly for legacy tests, before we got more robust with application-specific unit tests?

Personally, I've found that this series of tests, while annoying to maintain, is still useful...especially when doing schema changes that have unanticipated affects.

Not something that needs addressed to merge, just curious to hear your thoughts....would be good to be explicit about what we want the endpoint tests to accomplish.

File: usaspending_api/accounts/tests/test_federal_endpoints.py (hunk 8 above)
Comment (afrasier) [inline usaspending_api/accounts/tests/test_federal_endpoints.py]: I figured eventually, when we add sums on the federal account level by associated tas level data it would be useful. (Also largely leftover from a bit of copy and paste, so I can remove if needbe)

File: usaspending_api/accounts/views/common.py (hunk 5 above)
Comment (afrasier) [inline usaspending_api/accounts/views/common.py]: As you said it's kind of it's own thing, I'd rather leave it in common.py for now and move it when we give it a forever home. :)

File: usaspending_api/api_docs/markdown/data_dictionary.md (hunk 6 above)
Comment (afrasier) [inline usaspending_api/api_docs/markdown/data_dictionary.md]: Ack, that was supposed to be /autocomplete, but I goofed a copy and paste. Thanks for pointing it out

File: usaspending_api/data/testing_data/endpoint_testing_data.json (hunk 10 above)
Comment (afrasier) [inline usaspending_api/data/testing_data/endpoint_testing_data.json]: I think this was mainly a stop-gap for our previous lack of testing on each endpoint. I'd love to migrate these out into their own `test_whatever_endpoint.py` files to follow normal testing conventions. As it is, I've been trying to add new all tests to their own test file, and just keep this one updated for any previous tests. (This way we can also reduce reliance on "endpoint_fixture_db" and use our model generator for test data)
